File: sharedModule/src/main/java/ru/nekit/android/qls/setupWizard/BaseSetupWizardFragment.java
package ru.nekit.android.qls.setupWizard;

import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.StringRes;
import android.support.v4.app.Fragment;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;

import ru.nekit.android.shared.R;

import static android.view.View.INVISIBLE;
import static android.view.View.VISIBLE;

public abstract class BaseSetupWizardFragment extends Fragment {

    protected abstract void onSetupStart(@NonNull View view);

    @Override
    final public View onCreateView(LayoutInflater inflater, ViewGroup container,
                                   Bundle savedInstanceState) {
        super.onCreateView(inflater, container, savedInstanceState);
        View view = /*getLayoutId() == 0 ? getLayoutView() :*/
                inflater.inflate(getLayoutId(), container, false);
        setDefaultSettingsForTools();
        onSetupStart(view);
        return view;
    }

    protected BaseSetupWizard getSetupWizard() {
        return getSetupWizardHolder().getSetupWizard();
    }

    protected void showSetupWizardStep(@NonNull ISetupWizardStep step, Object... params) {
        getSetupWizardHolder().showSetupWizardStep(step, params);
    }

    private ISetupWizardHolder getSetupWizardHolder() {
        return (ISetupWizardHolder) getActivity();
    }

    protected void showNextSetupWizardStep() {
        showSetupWizardStep(getSetupWizard().getNextStep());
    }

    protected boolean addToBackStack() {
        return false;
    }

    private void setDefaultSettingsForTools() {
        setNextButtonText(R.string.label_next);
        setNextButtonVisibility(true);
        setAltButtonVisibility(false);
        setToolContainerVisibility(true);
    }

    public void setUnconditionedNextAction(boolean value) {
        getSetupWizardHolder().setUnconditionedNextAction(value);
    }

    final protected Button getNextButton() {
        return getSetupWizardHolder().getNextButton();
    }

    final protected Button getAltButton() {
        return getSetupWizardHolder().getAltButton();
    }

    final protected View getToolContainer() {
        return getSetupWizardHolder().getToolContainer();
    }

    final protected void setToolContainerVisibility(boolean visibility) {
        getToolContainer().setVisibility(visibility ? VISIBLE : INVISIBLE);
    }

    final protected void setNextButtonVisibility(boolean visibility) {
        getNextButton().setVisibility(visibility ? VISIBLE : INVISIBLE);
    }

    final protected void setAltButtonVisibility(boolean visibility) {
        getAltButton().setVisibility(visibility ? VISIBLE : INVISIBLE);
    }

    final protected void setNextButtonText(@StringRes int textResId) {
        if (textResId == 0) {
            textResId = R.string.label_next;
        }
        setNextButtonText(getString(textResId));
    }

    final protected void setNextButtonText(@NonNull String text) {
        getNextButton().setText(text);
    }

    final protected void setAltButtonText(@StringRes int textResId) {
        setAltButtonVisibility(true);
        getAltButton().setText(getString(textResId));
    }

    protected boolean nextAction() {
        return true;
    }

    protected void altAction() {
    }

    protected abstract int getLayoutId();

    /*protected View getLayoutView() {
        return null;
    }

    @Size(value = 4)
    public int[] getContainerMargins(@NonNull Resources resources) {
        int margin = resources.getDimensionPixelSize(R.dimen.base_semi_gap);
        return new int[]{
                margin,
                margin,
                margin,
                margin
        };
    }

    @Override
    public Animation onCreateAnimation(int transit, boolean enter, int nextAnim) {

       Animation anim = AnimationUtils.loadAnimation(getActivity(), nextAnim);

        anim.setAnimationListener(new Animation.AnimationListener() {

            @Override
            public void onAnimationStart(Animation animation) {

            }

            @Override
            public void onAnimationRepeat(Animation animation) {

            }

            @Override
            public void onAnimationEnd(Animation animation) {
                BaseSetupWizardFragment.this.onAnimationEnd();
            }
        });

        return anim;
    }

    protected void onAnimationEnd() {

    }*/

}

File: sharedModule/src/main/java/ru/nekit/android/qls/setupWizard/BaseSetupWizardFragment.kt
package ru.nekit.android.qls.setupWizard

import android.os.Bundle
import android.support.annotation.CallSuper
import android.support.annotation.LayoutRes
import android.support.annotation.StringRes
import android.support.v4.app.Fragment
import android.view.LayoutInflater
import android.view.View
import android.view.View.*
import android.view.ViewGroup
import android.widget.Button
import io.reactivex.Single
import io.reactivex.disposables.CompositeDisposable
import ru.nekit.android.shared.R
import ru.nekit.android.utils.IAutoDispose
import ru.nekit.android.utils.toSingle

abstract class BaseSetupWizardFragment : Fragment(), IAutoDispose {

    open val setupWizard: BaseSetupWizard
        get() = setupWizardHolder.setupWizard

    private val setupWizardHolder: ISetupWizardHolder
        get() = activity as ISetupWizardHolder

    private val nextButton: Button
        get() = setupWizardHolder.getNextButton()

    private val altButton: Button
        get() = setupWizardHolder.getAltButton()

    private val toolContainer: View
        get() = setupWizardHolder.getToolContainer()

    override var disposableMap: MutableMap<String, CompositeDisposable> = HashMap()

    @LayoutRes
    protected abstract fun getLayoutId(): Int

    protected abstract fun onSetupStart(view: View)

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        super.onCreateView(inflater, container, savedInstanceState)
        return inflater.inflate(getLayoutId(), container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        setDefaultSettingsForTools()
        onSetupStart(view)
    }

    protected fun showSetupWizardStep(step: ISetupWizardStep, vararg params: Any) {
        setupWizardHolder.showSetupWizardStep(step, *params)
    }

    protected open fun goNext() {
        setupWizardHolder.showNextSetupWizardStep()
    }

    open var title: Int
        get() = 0
        set(value) = setupWizardHolder.setTitle(value)

    open val addToBackStack: Boolean = false

    private fun setDefaultSettingsForTools() {
        nextButtonText(R.string.label_next)
        nextButtonVisibility(true)
        altButtonVisibility(false)
        toolContainerVisibility(true)
    }

    open var unconditionedNextAction: Boolean = false

    protected fun toolContainerVisibility(visibility: Boolean) {
        toolContainer.visibility = if (visibility) VISIBLE else GONE
    }

    protected fun nextButtonVisibility(visibility: Boolean) {
        nextButton.visibility = if (visibility) VISIBLE else INVISIBLE
    }

    protected fun altButtonVisibility(visibility: Boolean) {
        altButton.visibility = if (visibility) VISIBLE else INVISIBLE
    }

    protected fun nextButtonText(@StringRes textResId: Int) {
        var textResIdLocal = textResId
        if (textResId == 0) {
            textResIdLocal = R.string.label_next
        }
        nextButtonText(getString(textResIdLocal))
    }

    protected fun nextButtonText(text: String) {
        nextButton.text = text
    }

    protected fun altButtonText(@StringRes textResId: Int) {
        altButtonVisibility(true)
        altButton.text = getString(textResId)
    }

    open fun nextAction() = true.toSingle()

    open fun altAction() {
        if (addToBackStack) {
            goBack()
        }
    }

    protected fun goBack() = activity?.supportFragmentManager?.popBackStack()

    protected fun backAction(): Single<Boolean> = Single.fromCallable {
        goBack()
        false
    }

    @CallSuper
    override fun onDestroy() {
        dispose()
        super.onDestroy()
    }

}